src/Servers/Kestrel/Transport.Quic/src/Internal/QuicStreamContext.cs

```diff
@@ -0,0 +1,337 @@
+            {
+                _log.LogError(0, ex, $"Unexpected exception in {nameof(QuicStreamContext)}.{nameof(CancelConnectionClosedToken)}.");
+            }
+        }
+
+
+        private async Task DoSend()
+        {
+            Exception shutdownReason = null;
+            Exception unexpectedError = null;
+
+            try
+            {
+                await ProcessSends();
+            }
+            catch (QuicException ex)
+            {
+                shutdownReason = new ConnectionResetException(ex.Message, ex);
+            }
+            catch (Exception ex)
+            {
+                shutdownReason = ex;
+                unexpectedError = ex;
+                _log.ConnectionError(ConnectionId, unexpectedError);
+            }
+            finally
+            {
+                await ShutdownWrite(shutdownReason);
+
+                // Complete the output after disposing the stream
+                Output.Complete(unexpectedError);
+
+                // Cancel any pending flushes so that the input loop is un-paused
+                Input.CancelPendingFlush();
+            }
+        }
+
+        private async Task ProcessSends()
+        {
+            // Resolve `output` PipeReader via the IDuplexPipe interface prior to loop start for performance.
+            var output = Output;
+            while (true)
+            {
+                var result = await output.ReadAsync();
+
+                if (result.IsCanceled)
+                {
+                    break;
+                }
+
+                var buffer = result.Buffer;
+
+                var end = buffer.End;
+                var isCompleted = result.IsCompleted;
+                if (!buffer.IsEmpty)
+                {
+                    await _stream.WriteAsync(buffer, endStream: isCompleted);
+                }
+
+                output.AdvanceTo(end);
+
+                if (isCompleted)
+                {
+                    // Once the stream pipe is closed, shutdown the stream.
+                    break;
+                }
+            }
+        }
+
+        public override void Abort(ConnectionAbortedException abortReason)
+        {
+            _stream.AbortRead(_context.Options.AbortErrorCode);
+            _stream.AbortWrite(_context.Options.AbortErrorCode);
+
+            // Cancel ProcessSends loop after calling shutdown to ensure the correct _shutdownReason gets set.
+            Output.CancelPendingRead();
+        }
+
+        private async ValueTask ShutdownWrite(Exception shutdownReason)
+        {
+            // TODO synchronize this with Abort
+            _shutdownReason = shutdownReason ?? new ConnectionAbortedException("The Quic transport's send loop completed gracefully.");
+
+            _log.StreamShutdownWrite(ConnectionId, _shutdownReason);
+
+            try
+            {
+                // Try to gracefully close the stream
+                _stream.Shutdown();
+                await _stream.ShutdownWriteCompleted();
```

Comment:
Can you explain the difference between Shutdown() and ShutdownWriteCompleted() and why it's called in this order?

Reply:
Stream.Shutdown starts the Shutdown sequence, awaiting ShutdownWriteCompleted just awaits the task that would be returned from Shutdown.